Local-search label placement algorithm

WPFerg opened the pull request:
Uses a local search approach to place labels -- directly trying to resolve any possible collisions. It automatically terminates if it reaches an area of local minimum.

I've also edited the `minimum` util to allow parameters to be passed in to the scoring function (saves having to bind variables each time), but it can be switched back with no real differences

WPFerg commented:
Pushed up a slight refactor, bringing shared functions into their own utility file, making the scorer function a factory (code readability), and making the algorithm more stateless.

ColinEberhardt commented on src/layout/strategy/collision.js:
```diff
@@ -30,8 +30,36 @@ function collidePoint(data, pointIndex) {
+        }
+    });
+
+    return uniqueCollisions;
```
Isn't there a simpler way to compute this? Isn't it just a filter over the data where you check for any intesections?

ColinEberhardt commented on src/layout/strategy/local.js:
```diff
@@ -0,0 +1,83 @@
+
+            // Reached a local minimum?
+            var thisIterationScore = areaOfCollisions(iteratedData);
+            if (lastIterationScore === thisIterationScore) {
```
Where if `lastIterationScore` updated in this loop?

ColinEberhardt commented on src/layout/strategy/local.js:
```diff
@@ -0,0 +1,83 @@
+            // Reached a local minimum?
+            var thisIterationScore = areaOfCollisions(iteratedData);
+            if (lastIterationScore === thisIterationScore) {
+                return iteratedData;
```
Rather than an early-return, I think this would be more clearly expressed as a while loop, i.e. repeat while i < iterations AND score < previousScore

ColinEberhardt commented on src/layout/strategy/local.js:
```diff
@@ -0,0 +1,83 @@
+        // Try to resolve collisions from each node which has a collision
+        for (var i = 0; i < collidingPoints.length; i++) {
+
+            var pointIndex = collidingPoints[i];
```
Perhaps use Array.forEach to avoid the need to index into the array?

ColinEberhardt commented:
Just a few comments, I'm still a little unsure about the construction of the algorithm, I'll pop over and have a chat with you later.

ColinEberhardt commented:
@WPFerg - I've had a go at modifying the implementation. It's not 100% complete, but hopefully you get the idea:

https://github.com/ColinEberhardt/d3-financial-components/blob/WPFerg-labels-local/src/layout/strategy/local.js

This removes the need for the various utility methods within `collision.js` which I don't think were adding value, their outputs were a little confusing. I think you could also make more use of higher order functions, filter, map, reduce, forEach etc ...

I'd be happy to talk this over with you this morning.

WPFerg commented:
Thanks -- I've done the refactors you suggested yesterday and changed it to a more functional approach.

I've left the area calculations in `collision.js`, and refactored greedy accordingly (I've changed the tests due to the area of collisions calculation no longer doing self-collisions, which caused a change in output here due to the rectangles not all being the same size).

ColinEberhardt commented:
Many thanks for the updates - I've rebased and squashed, will merge shortly when Travis is happy
